### PR TITLE
Add records support to XmlDocCommentCompletion and ChangeSignature

### LINKED ISSUE #44558: ChangeSignature: should trigger in positional records
ChangeSignature is currently not offered.
Note: we first need to clarify whether those are parameter or property declarations (tracked by https://github.com/dotnet/roslyn/issues/44556).

image

Note that we offer ChangeSignature on other constructor parameters:
image

Relates to https://github.com/dotnet/roslyn/issues/40726 (test plan for "records")

### LINKED ISSUE #52738: Make sure XmlDocCommentCompletionProvider works with records
As a follow-up to https://github.com/dotnet/roslyn/pull/52737

Also when reordering parameters, such as what currently happens with methods:

https://github.com/dotnet/roslyn/blob/91571a3bb038e05e7bf2ab87510273a1017faed0/src/EditorFeatures/CSharpTest/ReorderParameters/ReorderParametersTests.cs#L427-L455

@jcouv, This may require parallel record structs work.

### PR DESCRIPTION
**Pull request by Youssef1313**
~~The change signature part is crashing. The assertion failure is far down to the compiler layer. I wasn't able to figure out how to fix that. Tagging @allisonchou for help here since the issue was assigned to you.~~ Figured out the reason. It's actually an existing bug (see https://github.com/dotnet/roslyn/pull/53089).

Fixes #44558
Fixes #52738

### TODOs

- [x] Integration tests in https://github.com/dotnet/roslyn/blob/main/src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests_XmlDoc.vb.
- [x] Add tests in https://github.com/dotnet/roslyn/blob/main/src/EditorFeatures/CSharpTest/ReorderParameters/ReorderParametersTests.cs and https://github.com/dotnet/roslyn/blob/main/src/EditorFeatures/CSharpTest/ChangeSignature/ReorderParametersTests.cs
- [x] Add tests in https://github.com/dotnet/roslyn/blob/main/src/EditorFeatures/CSharpTest/ChangeSignature/AddParameterTests.Cascading.cs for inheritance scenarios

### CODE REVIEW

File: src/Features/Core/Portable/ChangeSignature/AbstractChangeSignatureService.cs [hunk 1]
```diff
@@ -124,6 +124,19 @@ public async Task<ImmutableArray<ChangeSignatureCodeAction>> GetChangeSignatureC
                 {
                     symbol = typeSymbol.DelegateInvokeMethod;
                 }
+                else if (typeSymbol.IsRecord)
+                {
+                    // A bit hacky to determine the parameters of primary constructor associated with a given record.
+                    // TODO: record structs.
```
Comment (Youssef1313) [inline src/Features/Core/Portable/ChangeSignature/AbstractChangeSignatureService.cs:130]: @jcouv FYI to track this.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs [hunk 2]
```diff
@@ -364,13 +364,33 @@ public static bool IsParams([NotNullWhen(returnValue: true)] this ISymbol? symbo
...
+                // TODO: record structs.
+                var primaryConstructor = recordSymbol.InstanceConstructors.FirstOrDefault(
+                    c => c.DeclaringSyntaxReferences.FirstOrDefault()?.GetSyntax()?.RawKind == RecordDeclarationRawKind);
+
+                if (primaryConstructor is null)
+                {
+                    return ImmutableArray<IParameterSymbol>.Empty;
+                }
+
+                return primaryConstructor.Parameters;
```
Comment (Youssef1313) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs:390]: Having primary constructor with `AssociatedSymbol` equals to the record symbol would save this ugly workaround :smile:

@jcouv @jaredpar Is this something you want to do in the compiler side?

**Comment by Youssef1313** [comment]
CI is stuck:

image


Closing and reopening to retrigger run.

**Comment by Youssef1313** [comment]
Closing and reopening:

```
C:\Users\VssAdministrator\.nuget\packages\microsoft.dotnet.helix.sdk\6.0.0-beta.21167.3\tools\azure-pipelines\AzurePipelines.MultiQueue.targets(30,5): error : System.Net.Http.HttpRequestException: Response status code does not indicate success: 500 (Internal Server Error). [D:\a\1\s\helix-tmp.csproj]
C:\Users\VssAdministrator\.nuget\packages\microsoft.dotnet.helix.sdk\6.0.0-beta.21167.3\tools\azure-pipelines\AzurePipelines.MultiQueue.targets(30,5): error :    at System.Net.Http.HttpResponseMessage.EnsureSuccessStatusCode() [D:\a\1\s\helix-tmp.csproj]
C:\Users\VssAdministrator\.nuget\packages\microsoft.dotnet.helix.sdk\6.0.0-beta.21167.3\tools\azure-pipelines\AzurePipelines.MultiQueue.targets(30,5): error :    at Microsoft.DotNet.Helix.Sdk.CreateTestsForWorkItems.<>c__DisplayClass5_0.<<AttachResultFileToTestResultAsync>b__0>d.MoveNext() in /_/src/Microsoft.DotNet.Helix/Sdk/CreateFailedTestsForFailedWorkItems.cs:line 73 [D:\a\1\s\helix-tmp.csproj]
C:\Users\VssAdministrator\.nuget\packages\microsoft.dotnet.helix.sdk\6.0.0-beta.21167.3\tools\azure-pipelines\AzurePipelines.MultiQueue.targets(30,5): error : --- End of stack trace from previous location --- [D:\a\1\s\helix-tmp.csproj]
C:\Users\VssAdministrator\.nuget\packages\microsoft.dotnet.helix.sdk\6.0.0-beta.21167.3\tools\azure-pipelines\AzurePipelines.MultiQueue.targets(30,5): error :    at Microsoft.DotNet.Helix.AzureDevOps.AzureDevOpsTask.<>c__DisplayClass13_0.<<RetryAsync>b__0>d.MoveNext() in /_/src/Microsoft.DotNet.Helix/Sdk/AzureDevOpsTask.cs:line 91 [D:\a\1\s\helix-tmp.csproj]
C:\Users\VssAdministrator\.nuget\packages\microsoft.dotnet.helix.sdk\6.0.0-beta.21167.3\tools\azure-pipelines\AzurePipelines.MultiQueue.targets(30,5): error : --- End of stack trace from previous location --- [D:\a\1\s\helix-tmp.csproj]
C:\Users\VssAdministrator\.nuget\packages\microsoft.dotnet.helix.sdk\6.0.0-beta.21167.3\tools\azure-pipelines\AzurePipelines.MultiQueue.targets(30,5): error :    at Microsoft.DotNet.Helix.AzureDevOps.AzureDevOpsTask.RetryAsync[T](Func`1 function, Action`1 logRetry, Func`2 isRetryable, CancellationToken cancellationToken) in /_/src/Microsoft.DotNet.Helix/Sdk/AzureDevOpsTask.cs:line 206 [D:\a\1\s\helix-tmp.csproj]
C:\Users\VssAdministrator\.nuget\packages\microsoft.dotnet.helix.sdk\6.0.0-beta.21167.3\tools\azure-pipelines\AzurePipelines.MultiQueue.targets(30,5): error :    at Microsoft.DotNet.Helix.AzureDevOps.AzureDevOpsTask.RetryAsync(Func`1 function) in /_/src/Microsoft.DotNet.Helix/Sdk/AzureDevOpsTask.cs:line 86 [D:\a\1\s\helix-tmp.csproj]
```

**Comment by Youssef1313** [comment]
```
C:\Users\VssAdministrator\.nuget\packages\microsoft.dotnet.helix.sdk\6.0.0-beta.21167.3\tools\azure-pipelines\AzurePipelines.MultiQueue.targets(30,5): error : System.Net.Http.HttpRequestException: Response status code does not indicate success: 500 (Internal Server Error). [D:\a\1\s\helix-tmp.csproj]
```

Is this a known issue? Re-run CI for the third time with the same error.

File: src/Features/Core/Portable/ChangeSignature/AbstractChangeSignatureService.cs [hunk 3]
```diff
@@ -124,6 +124,19 @@ public async Task<ImmutableArray<ChangeSignatureCodeAction>> GetChangeSignatureC
...
+                    const int RecordDeclarationRawKind = 9063;
+                    var potentialPrimaryCtor = typeSymbol.InstanceConstructors.FirstOrDefault(
```
Comment (allisonchou) [inline src/Features/Core/Portable/ChangeSignature/AbstractChangeSignatureService.cs:132]: Hmm, I'm a little worried about hardcoding the raw kind here. I almost think we should wait on merging until we have a proper workaround that doesn't involve hardcoding. @CyrusNajmabadi do you have any thoughts?

File: src/Features/CSharp/Portable/ChangeSignature/CSharpChangeSignatureService.cs [hunk 4]
```diff
@@ -42,7 +42,9 @@ internal sealed class CSharpChangeSignatureService : AbstractChangeSignatureServ
             SyntaxKind.DelegateDeclaration,
             SyntaxKind.SimpleLambdaExpression,
             SyntaxKind.ParenthesizedLambdaExpression,
-            SyntaxKind.LocalFunctionStatement);
+            SyntaxKind.LocalFunctionStatement,
+            // TODO: Record structs
```
[the remark this answers is not in the document]
Comment (Youssef1313) [reply src/Features/CSharp/Portable/ChangeSignature/CSharpChangeSignatureService.cs:46]: @allisonchou It should fall into the existing test plan issue for record structs (https://github.com/dotnet/roslyn/issues/51199). Tagging @jcouv to edit the issue

File: src/Features/Core/Portable/ChangeSignature/AbstractChangeSignatureService.cs (hunk 3 above)
Comment (Youssef1313) [reply src/Features/Core/Portable/ChangeSignature/AbstractChangeSignatureService.cs:132]: I think https://github.com/dotnet/roslyn/issues/53092 may not be very soon 😕

File: src/Features/CSharp/Portable/ChangeSignature/CSharpChangeSignatureService.cs [hunk 5]
```diff
@@ -286,6 +291,13 @@ private static bool InSymbolHeader(SyntaxNode matchingNode, int position)
                 return method.WithParameterList(method.ParameterList.WithParameters(updatedParameters).WithAdditionalAnnotations(changeSignatureFormattingAnnotation));
             }
 
+            // TODO: Record structs
+            if (updatedNode.IsKind(SyntaxKind.RecordDeclaration, out RecordDeclarationSyntax? record) && record.ParameterList is not null)
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/ChangeSignature/CSharpChangeSignatureService.cs:295]: just don't do a kind check, and do a syntax-check, and you'll be all good here.  
```suggestion
            if (updatedNode is RecordDeclarationSyntax { ParameterList: not null })
```

File: src/Features/Core/Portable/ChangeSignature/AbstractChangeSignatureService.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/ChangeSignature/AbstractChangeSignatureService.cs:132]: we're in an abstract type, just provide a helper that C# implements, going back to source to answer this.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs [hunk 6]
```diff
@@ -364,13 +364,33 @@ public static bool IsParams([NotNullWhen(returnValue: true)] this ISymbol? symbo
...
 
         public static ImmutableArray<IParameterSymbol> GetParameters(this ISymbol? symbol)
-            => symbol switch
+        {
+            return symbol switch
             {
                 IMethodSymbol m => m.Parameters,
-                IPropertySymbol nt => nt.Parameters,
+                IPropertySymbol p => p.Parameters,
+                INamedTypeSymbol nt when nt.IsRecord => GetRecordParameters(nt),
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs:372]: yeah, i do not like this *at all*.  i'd say the caller needs to potentially deal with records.

File: src/Features/CSharp/Portable/ChangeSignature/CSharpChangeSignatureService.cs (hunk 5 above)
Comment (Youssef1313) [reply src/Features/CSharp/Portable/ChangeSignature/CSharpChangeSignatureService.cs:295]: image

@CyrusNajmabadi Nullable analysis doesn't seem to handle patterns correctly.

File: src/Features/Core/Portable/ChangeSignature/AbstractChangeSignatureService.cs [hunk 7]
```diff
@@ -126,15 +126,9 @@ public async Task<ImmutableArray<ChangeSignatureCodeAction>> GetChangeSignatureC
...
-                    // TODO: record structs.
-                    const int RecordDeclarationRawKind = 9063;
-                    var potentialPrimaryCtor = typeSymbol.InstanceConstructors.FirstOrDefault(
-                        c => c.DeclaringSyntaxReferences.FirstOrDefault()?.GetSyntax()?.RawKind == RecordDeclarationRawKind);
-
-                    if (potentialPrimaryCtor is not null)
+                    if (TryGetRecordPrimaryConstructor(typeSymbol, out var primaryConstructor))
                     {
-                        symbol = potentialPrimaryCtor;
+                        symbol = primaryConstructor;
```
Comment (Youssef1313) [inline src/Features/Core/Portable/ChangeSignature/AbstractChangeSignatureService.cs]: @CyrusNajmabadi I forgot to add `[NotNullWhen(true)]` to the out parameter, but... There is NO warning here, which is weird. `symbol` is here assigned a "maybe null" value, and is being dereferenced a few lines later.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs (hunk 6 above)
Comment (Youssef1313) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions.cs]: @CyrusNajmabadi Done. However, it became a little bit more complicated.

**Comment by Youssef1313** [comment]
Pinging @CyrusNajmabadi and @allisonchou for review.

**Review by allisonchou — COMMENTED** [review]
Looking really good! Just have a few small comments/questions.

File: src/Features/CSharp/Portable/ChangeSignature/CSharpChangeSignatureService.cs [hunk 8]
```diff
@@ -875,5 +889,35 @@ protected override bool SupportsOptionalAndParamsArrayParametersSimultaneously()
...
+        {
+            Debug.Assert(typeSymbol.IsRecord);
+            primaryConstructor = typeSymbol.InstanceConstructors.FirstOrDefault(
+                c => c.DeclaringSyntaxReferences.FirstOrDefault()?.GetSyntax() is RecordDeclarationSyntax);
+            return primaryConstructor is not null;
+        }
+
+        protected override ImmutableArray<IParameterSymbol> GetParameters(ISymbol declarationSymbol)
+        {
+            var declaredParameters = declarationSymbol.GetParameters();
```
[the remark this answers is not in the document]
Comment (Youssef1313) [reply src/Features/CSharp/Portable/ChangeSignature/CSharpChangeSignatureService.cs:903]: @allisonchou I'm using it inside the if body, outside the if body, and in the condition.

File: src/Features/CSharp/Portable/ChangeSignature/CSharpChangeSignatureService.cs [hunk 9]
```diff
@@ -265,12 +270,13 @@ private static bool InSymbolHeader(SyntaxNode matchingNode, int position)
             CancellationToken cancellationToken)
         {
             var updatedNode = potentiallyUpdatedNode as CSharpSyntaxNode;
-
```
[the remark this answers is not in the document]
Comment (Youssef1313) [reply src/Features/CSharp/Portable/ChangeSignature/CSharpChangeSignatureService.cs:268]: This was unintentionally deleted.

File: src/Features/CSharp/Portable/ChangeSignature/CSharpChangeSignatureService.cs [hunk 10]
```diff
@@ -875,5 +889,35 @@ protected override bool SupportsOptionalAndParamsArrayParametersSimultaneously()
...
+        protected override ImmutableArray<IParameterSymbol> GetParameters(ISymbol declarationSymbol)
+        {
+            var declaredParameters = declarationSymbol.GetParameters();
+            if (declarationSymbol is INamedTypeSymbol { IsRecord: true } recordSymbol)
+            {
+                Debug.Assert(declaredParameters.IsDefaultOrEmpty, "If GetParameters extension handles record, we can remove the handling here.");
+                // A bit hacky to determine the parameters of primary constructor associated with a given record.
+                // Simplifying is tracked by: https://github.com/dotnet/roslyn/issues/53092.
+                // Note: When the issue is handled, we can remove the logic here and handle things in GetParameters. BUT
+                // make sure none of the other callers doesn't need records to be handled.
```
[the remark this answers is not in the document]
Comment (Youssef1313) [reply src/Features/CSharp/Portable/ChangeSignature/CSharpChangeSignatureService.cs]: Updated the comment.

File: src/Features/CSharp/Portable/ChangeSignature/CSharpChangeSignatureService.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (allisonchou) [reply src/Features/CSharp/Portable/ChangeSignature/CSharpChangeSignatureService.cs:904]: whoops, wasn't paying attention closely enough. Gotcha, makes sense!

**Review by allisonchou — APPROVED** [review]
Thanks! This LGTM. 😄

**Comment by allisonchou** [comment]
@Youssef1313 It looks like there's some check failures, could you take a look?

**Comment by Youssef1313** [comment]
@CyrusNajmabadi Can you review please? Thanks!

File: src/Features/CSharp/Portable/ChangeSignature/CSharpChangeSignatureService.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/ChangeSignature/CSharpChangeSignatureService.cs:47]: we have record structs now right?  so can we just add that in this PR?

File: src/Features/CSharp/Portable/ChangeSignature/CSharpChangeSignatureService.cs [hunk 11]
```diff
@@ -875,5 +890,36 @@ protected override bool SupportsOptionalAndParamsArrayParametersSimultaneously()
...
+            if (declarationSymbol is INamedTypeSymbol { IsRecord: true } recordSymbol)
+            {
+                Debug.Assert(declaredParameters.IsDefaultOrEmpty, "If GetParameters extension handles record, we can remove the handling here.");
+                // A bit hacky to determine the parameters of primary constructor associated with a given record.
+                // Simplifying is tracked by: https://github.com/dotnet/roslyn/issues/53092.
+                // Note: When the issue is handled, we can remove the logic here and handle things in GetParameters extension. BUT
+                // if GetParameters extension method gets updated to handle records, we need to test EVERY usage
+                // of the extension method and make sure the change is applicable to all these usages.
+                var primaryConstructor = recordSymbol.InstanceConstructors.FirstOrDefault(
+                    c => c.DeclaringSyntaxReferences.FirstOrDefault()?.GetSyntax() is RecordDeclarationSyntax);
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/ChangeSignature/CSharpChangeSignatureService.cs:914]: consider extracting helper for this (you do the same work higher up).

File: src/Features/CSharp/Portable/Completion/CompletionProviders/XmlDocCommentCompletionProvider.cs [hunk 12]
```diff
@@ -365,6 +366,29 @@ private string GetAttributeValue(XmlAttributeSyntax attribute)
...
+            if (declarationSymbol is INamedTypeSymbol { IsRecord: true } recordSymbol)
+            {
+                Debug.Assert(declaredParameters.IsDefaultOrEmpty, "If GetParameters extension handles record, we can remove the handling here.");
+                // A bit hacky to determine the parameters of primary constructor associated with a given record.
+                // Simplifying is tracked by: https://github.com/dotnet/roslyn/issues/53092.
+                // Note: When the issue is handled, we can remove the logic here and handle things in GetParameters extension. BUT
+                // if GetParameters extension method gets updated to handle records, we need to test EVERY usage
+                // of the extension method and make sure the change is applicable to all these usages.
+                var primaryConstructor = recordSymbol.InstanceConstructors.FirstOrDefault(
+                    c => c.DeclaringSyntaxReferences.FirstOrDefault()?.GetSyntax() is RecordDeclarationSyntax);
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/XmlDocCommentCompletionProvider.cs:381]: used again.  consider an extension now.

File: src/Features/Core/Portable/ChangeSignature/AbstractChangeSignatureService.cs [hunk 13]
```diff
@@ -124,6 +132,13 @@ public async Task<ImmutableArray<ChangeSignatureCodeAction>> GetChangeSignatureC
                 {
                     symbol = typeSymbol.DelegateInvokeMethod;
                 }
+                else if (typeSymbol.IsRecord)
+                {
+                    if (TryGetRecordPrimaryConstructor(typeSymbol, out var primaryConstructor))
+                    {
+                        symbol = primaryConstructor;
+                    }
+                }
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/ChangeSignature/AbstractChangeSignatureService.cs:141]:
```suggestion
                else if (typeSymbol.IsRecord && TryGetRecordPrimaryConstructor(typeSymbol, out var primaryConstructor))
                    {
                        symbol = primaryConstructor;
                    }
```

**Review by CyrusNajmabadi — APPROVED** [review]
lgtm.  let me know if you want to add the support for record structs.

File: src/Features/CSharp/Portable/Completion/CompletionProviders/XmlDocCommentCompletionProvider.cs [hunk 14]
```diff
@@ -369,21 +369,10 @@ private string GetAttributeValue(XmlAttributeSyntax attribute)
         protected override ImmutableArray<IParameterSymbol> GetParameters(ISymbol declarationSymbol)
         {
             var declaredParameters = declarationSymbol.GetParameters();
-            if (declarationSymbol is INamedTypeSymbol { IsRecord: true } recordSymbol)
+            if (declarationSymbol is INamedTypeSymbol { IsRecord: true } recordSymbol &&
+                recordSymbol.TryGetRecordPrimaryConstructor(out var primaryConstructor))
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/XmlDocCommentCompletionProvider.cs:373]: You could move the IsRecord check into TryGetPrimaryConstructor

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
Thanks!

**Pull request merged**